ddtrace/internal/products.py

```diff
@@ -0,0 +1,178 @@
+import atexit
+from collections import defaultdict
+from collections import deque
+import sys
+import typing as t
+
+from ddtrace.internal import forksafe
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.uwsgi import check_uwsgi
+from ddtrace.internal.uwsgi import uWSGIConfigError
+from ddtrace.internal.uwsgi import uWSGIMasterProcess
+
+
+log = get_logger(__name__)
+
+if sys.version_info < (3, 10):
+    from importlib_metadata import entry_points
+else:
+    from importlib.metadata import entry_points
+
+try:
+    from typing import Protocol  # noqa:F401
+except ImportError:
+    from typing_extensions import Protocol  # type: ignore[assignment]
+
+
+class Product(Protocol):
+    requires: t.List[str]
+
+    def post_preload(self) -> None:
+        ...
+
+    def start(self) -> None:
+        ...
+
+    def restart(self, join: bool = False) -> None:
+        ...
+
+    def stop(self, join: bool = False) -> None:
+        ...
+
+    def at_exit(self, join: bool = False) -> None:
+        ...
+
+
+class ProductManager:
+    __products__: t.Dict[str, Product] = {}  # All discovered products
+
+    def __init__(self) -> None:
+        self._products: t.Optional[t.List[t.Tuple[str, Product]]] = None  # Topologically sorted products
+
+        for product_plugin in entry_points(group="ddtrace.products"):
+            name = product_plugin.name
+            log.debug("Discovered product plugin '%s'", name)
+
+            # Load the product protocol object
+            try:
+                product: Product = product_plugin.load()
+            except Exception:
+                log.exception("Failed to load product plugin '%s'", name)
+                continue
+
+            log.debug("Product plugin '%s' loaded successfully", name)
+
+            self.__products__[name] = product
+
+    def _sort_products(self) -> t.List[t.Tuple[str, Product]]:
+        # Data structures for topological sorting
+        q: t.Deque[str] = deque()  # Queue of products with no dependencies
+        g = defaultdict(list)  # Graph of dependencies
```

Comment:
nit: Would it make sense to make this a `set` instead? So that the `remove()` later on is less expensive? Probably not a huge deal since this will only run in when the service starts up

Reply:
Ah good point. I think we actually need to do this for `f`, since we are just iterating over the values of `g`